alonzo-ty/src/lib.rs

```diff
@@ -0,0 +1,36 @@
+use std::ops::Deref;
+
+pub trait BaseTy: Clone {
+    type Value: Clone;
+}
+
+#[derive(Clone)]
+pub enum Ty<T: BaseTy> {
+    Base(T),
+    Func(Box<Self>, Box<Self>),
```

Comment:
Hmm, I think we might want to specialize this type between `mir` and `lir`. For example, we probably want to include information about the function payload in `lir`.

Reply:
And, in general, we might be able to do more detailed lowering for other types that might warrant having definitions of `Ty` that are specific to `alonzo_lir` and `alonzo_hir`.

Reply:
But should the function payload info be stored in the type of the object? What else could be different between these two?

Reply:
If we're lowering things from `mir` to `lir`, we will probably want to do other sort of lowering such as turning, say, lists of a known size into static arrays with the length as part of the type.